## Introduction

YOLOv5 (You Only Look Once, version 5) is a top-tier object detection model developed by Ultralytics. It excels in real-time object detection for images and videos. This guide will help you set up YOLOv5 on your local system, including cloning the repository, installing necessary dependencies, and confirming your PyTorch and GPU configurations. By the end, you'll be ready to train or test models using YOLOv5.

## 1. Clone the YOLOv5 Repository


In [1]:
!git clone https://github.com/ultralytics/yolov5  # clone
%cd yolov5
%pip install -qr requirements.txt comet_ml  # install

import torch
import utils
display = utils.notebook_init()  # checks

YOLOv5 🚀 v7.0-365-g12b577c8 Python-3.10.12 torch-2.3.1+cu121 CUDA:0 (Tesla T4, 15102MiB)


Setup complete ✅ (2 CPUs, 12.7 GB RAM, 30.2/112.6 GB disk)


In [4]:
!git config --global user.email "Ekam.jyot.16@gmail.com"
!git config --global user.name "Ekamjyot1234@#"

In [5]:
!git add .
!git commit -m "First Commit"


[master (root-commit) d49f9d6] First Commit
 147 files changed, 30009 insertions(+)
 create mode 100644 .dockerignore
 create mode 100644 .gitattributes
 create mode 100644 .github/ISSUE_TEMPLATE/bug-report.yml
 create mode 100644 .github/ISSUE_TEMPLATE/config.yml
 create mode 100644 .github/ISSUE_TEMPLATE/feature-request.yml
 create mode 100644 .github/ISSUE_TEMPLATE/question.yml
 create mode 100644 .github/dependabot.yml
 create mode 100644 .github/workflows/ci-testing.yml
 create mode 100644 .github/workflows/cla.yml
 create mode 100644 .github/workflows/codeql-analysis.yml
 create mode 100644 .github/workflows/docker.yml
 create mode 100644 .github/workflows/format.yml
 create mode 100644 .github/workflows/greetings.yml
 create mode 100644 .github/workflows/links.yml
 create mode 100644 .github/workflows/merge-main-into-prs.yml
 create mode 100644 .github/workflows/stale.yml
 create mode 100755 .gitignore
 create mode 100644 CITATION.cff
 create mode 100644 CONTRIBUTING.md
 create 

In [6]:
!git remote add origin https://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@github.com/ekamjyot1515/AIDI1002_FinalAssignment.git


In [7]:
!git push origin master



Enumerating objects: 167, done.
Counting objects: 100% (167/167), done.
Delta compression using up to 2 threads
Compressing objects: 100% (166/166), done.
Writing objects: 100% (167/167), 992.35 KiB | 11.41 MiB/s, done.
Total 167 (delta 43), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (43/43), done.
To https://github.com/ekamjyot1515/AIDI1002_FinalAssignment.git
 * [new branch]      master -> master


# 1. Detect

`detect.py` runs YOLOv5 inference on a variety of sources, downloading models automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases), and saving results to `runs/detect`. Example inference sources are:

```shell
python detect.py --source 0  # webcam
                          img.jpg  # image
                          vid.mp4  # video
                          screen  # screenshot
                          path/  # directory
                         'path/*.jpg'  # glob
                         'https://youtu.be/LNwODJXcvt4'  # YouTube
                         'rtsp://example.com/media.mp4'  # RTSP, RTMP, HTTP stream
```


In [8]:
!python detect.py --weights yolov5s.pt --img 640 --conf 0.25 --source data/images
# display.Image(filename='runs/detect/exp/zidane.jpg', width=600)

detect: weights=['yolov5s.pt'], source=data/images, data=data/coco128.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_format=0, save_csv=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
YOLOv5 🚀 d49f9d6 Python-3.10.12 torch-2.3.1+cu121 CUDA:0 (Tesla T4, 15102MiB)

100% 14.1M/14.1M [00:00<00:00, 460MB/s]

Fusing layers... 
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients, 16.4 GFLOPs
WARNING ⚠️ NMS time limit 0.550s exceeded
image 1/2 /content/yolov5/data/images/bus.jpg: 640x480 4 persons, 1 bus, 50.7ms
image 2/2 /content/yolov5/data/images/zidane.jpg: 384x640 2 persons, 2 ties, 52.7ms
Speed: 0.6ms pre-process, 51.7ms inference, 459.5ms NMS per image at shape (1, 3, 640, 640)
Results save

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<img align="left" src="https://user-images.githubusercontent.com/26833433/127574988-6a558aa1-d268-44b9-bf6b-62d4c605cc72.jpg" width="600">

# 2. Validate
Validate a model's accuracy on the [COCO](https://cocodataset.org/#home) dataset's `val` or `test` splits. Models are downloaded automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases). To show results by class use the `--verbose` flag.

In [9]:
# Download COCO val
torch.hub.download_url_to_file('https://github.com/ultralytics/assets/releases/download/v0.0.0/coco2017val.zip', 'tmp.zip')  # download (780M - 5000 images)
!unzip -q tmp.zip -d ../datasets && rm tmp.zip  # unzip

100%|██████████| 780M/780M [00:22<00:00, 37.0MB/s]


In [10]:
# Validate YOLOv5s on COCO val
!python val.py --weights yolov5s.pt --data coco.yaml --img 640 --half

val: data=/content/yolov5/data/coco.yaml, weights=['yolov5s.pt'], batch_size=32, imgsz=640, conf_thres=0.001, iou_thres=0.6, max_det=300, task=val, device=, workers=8, single_cls=False, augment=False, verbose=False, save_txt=False, save_hybrid=False, save_conf=False, save_json=True, project=runs/val, name=exp, exist_ok=False, half=True, dnn=False
YOLOv5 🚀 d49f9d6 Python-3.10.12 torch-2.3.1+cu121 CUDA:0 (Tesla T4, 15102MiB)

Fusing layers... 
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients, 16.4 GFLOPs
val: Scanning /content/datasets/coco/val2017... 4952 images, 48 backgrounds, 0 corrupt: 100% 5000/5000 [00:04<00:00, 1020.01it/s]
val: New cache created: /content/datasets/coco/val2017.cache
                 Class     Images  Instances          P          R      mAP50   mAP50-95: 100% 157/157 [01:30<00:00,  1.73it/s]
                   all       5000      36335      0.671      0.519      0.566      0.371
Speed: 0.1ms pre-process, 4.1ms inference, 2.6ms NMS per image at shape 

# 3. Train

<p align=""><a href="https://ultralytics.com/hub"><img width="1000" src="https://github.com/ultralytics/assets/raw/main/im/integrations-loop.png"/></a></p>
Close the active learning loop by sampling images from your inference conditions with the `roboflow` pip package
<br><br>

Train a YOLOv5s model on the [COCO128](https://www.kaggle.com/ultralytics/coco128) dataset with `--data coco128.yaml`, starting from pretrained `--weights yolov5s.pt`, or from randomly initialized `--weights '' --cfg yolov5s.yaml`.

- **Pretrained [Models](https://github.com/ultralytics/yolov5/tree/master/models)** are downloaded
automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases)
- **[Datasets](https://github.com/ultralytics/yolov5/tree/master/data)** available for autodownload include: [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [COCO128](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [Argoverse](https://github.com/ultralytics/yolov5/blob/master/data/Argoverse.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml), [xView](https://github.com/ultralytics/yolov5/blob/master/data/xView.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [SKU-110K](https://github.com/ultralytics/yolov5/blob/master/data/SKU-110K.yaml).
- **Training Results** are saved to `runs/train/` with incrementing run directories, i.e. `runs/train/exp2`, `runs/train/exp3` etc.
<br>

A **Mosaic Dataloader** is used for training which combines 4 images into 1 mosaic.

## Label a dataset on Roboflow (optional)

[Roboflow](https://roboflow.com/?ref=ultralytics) enables you to easily **organize, label, and prepare** a high quality dataset with your own custom data. Roboflow also makes it easy to establish an active learning pipeline, collaborate with your team on dataset improvement, and integrate directly into your model building workflow with the `roboflow` pip package.

In [11]:
#@title Select YOLOv5 🚀 logger {run: 'auto'}
logger = 'Comet' #@param ['Comet', 'ClearML', 'TensorBoard']

if logger == 'Comet':
  %pip install -q comet_ml
  import comet_ml; comet_ml.init()
elif logger == 'ClearML':
  %pip install -q clearml
  import clearml; clearml.browser_login()
elif logger == 'TensorBoard':
  %load_ext tensorboard
  %tensorboard --logdir runs/train

COMET WARNING: comet_ml.init() is deprecated and will be removed soon. Please use comet_ml.login()


Please paste your Comet API key from https://www.comet.com/api/my/settings/
(api key may not show as you type)
Comet API key: ··········


COMET INFO: Valid Comet API Key saved in /root/.comet.config (set COMET_CONFIG to change where it is saved).


In [12]:
# Train YOLOv5s on COCO128 for 3 epochs
!python train.py --img 640 --batch 16 --epochs 3 --data coco128.yaml --weights yolov5s.pt --cache

2024-09-07 05:30:06.449064: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-09-07 05:30:06.449138: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-09-07 05:30:06.623218: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
train: weights=yolov5s.pt, cfg=, data=coco128.yaml, hyp=data/hyps/hyp.scratch-low.yaml, epochs=3, batch_size=16, imgsz=640, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, noplots=False, evolve=None, evolve_population=data/hyps, resume_evolve=None, bucket=, cache=ram, image_weights=False, device=, multi_scale=False, single_cls=False, opti

# 4. Visualize

## Comet Logging and Visualization 🌟 NEW

[Comet](https://www.comet.com/site/lp/yolov5-with-comet/?utm_source=yolov5&utm_medium=partner&utm_campaign=partner_yolov5_2022&utm_content=yolov5_colab) is now fully integrated with YOLOv5. Track and visualize model metrics in real time, save your hyperparameters, datasets, and model checkpoints, and visualize your model predictions with [Comet Custom Panels](https://www.comet.com/docs/v2/guides/comet-dashboard/code-panels/about-panels/?utm_source=yolov5&utm_medium=partner&utm_campaign=partner_yolov5_2022&utm_content=yolov5_colab)! Comet makes sure you never lose track of your work and makes it easy to share results and collaborate across teams of all sizes!

Getting started is easy:
```shell
pip install comet_ml  # 1. install
export COMET_API_KEY=<Your API Key>  # 2. paste API key
python train.py --img 640 --epochs 3 --data coco128.yaml --weights yolov5s.pt  # 3. train
```
To learn more about all of the supported Comet features for this integration, check out the [Comet Tutorial](https://docs.ultralytics.com/yolov5/tutorials/comet_logging_integration). If you'd like to learn more about Comet, head over to our [documentation](https://www.comet.com/docs/v2/?utm_source=yolov5&utm_medium=partner&utm_campaign=partner_yolov5_2022&utm_content=yolov5_colab). Get started by trying out the Comet Colab Notebook:
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1RG0WOQyxlDlo5Km8GogJpIEJlg_5lyYO?usp=sharing)

<a href="https://bit.ly/yolov5-readme-comet2">
<img alt="Comet Dashboard" src="https://user-images.githubusercontent.com/26833433/202851203-164e94e1-2238-46dd-91f8-de020e9d6b41.png" width="1280"/></a>

In [14]:
!pip install comet_ml  # 1. install
!export COMET_API_KEY=0bPrFkFB552cKtF7UJxQoJ4jg  # 2. paste API key
!python train.py --img 640 --epochs 3 --data coco128.yaml --weights yolov5s.pt  # 3. train

2024-09-07 05:32:00.458314: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-09-07 05:32:00.458369: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-09-07 05:32:00.459870: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
train: weights=yolov5s.pt, cfg=, data=coco128.yaml, hyp=data/hyps/hyp.scratch-low.yaml, epochs=3, batch_size=16, imgsz=640, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, noplots=False, evolve=None, evolve_population=data/hyps, resume_evolve=None, bucket=, cache=None, image_weights=False, device=, multi_scale=False, single_cls=False, opt

# **Significant Contribution: Testing the Model on a New Dataset**

As a significant contribution to the methodology of the selected research paper, we have tested the model on a new dataset, **MNIST**, to evaluate its effectiveness in a different context.

#### Steps Involved:

1. **Importing Necessary Libraries**:
   - Added required TensorFlow and Keras libraries to handle the MNIST dataset.

2. **Loading and Preprocessing the MNIST Dataset**:
   - Loaded the MNIST dataset and normalized the pixel values to be between 0 and 1.
   - Reshaped the images to include a channel dimension and applied one-hot encoding to the labels for compatibility with the model.

3. **Modifying the Model Architecture**:
   - Adjusted the input shape in the model to accommodate the MNIST dataset (28x28x1).
   - Configured the output layer to have 10 classes, corresponding to the 10 digits in MNIST.

4. **Training the Model on MNIST**:
   - Trained the model for 10 epochs using the MNIST training set.
   - Used the validation set to monitor the model's performance during training.

5. **Evaluating the Model on the Test Set**:
   - Evaluated the trained model on the MNIST test set and recorded the test accuracy.


In [20]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Conv2D, Flatten, MaxPooling2D
from tensorflow.keras.utils import to_categorical

# Load MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Normalize pixel values to be between 0 and 1
x_train, x_test = x_train / 255.0, x_test / 255.0

# Reshape data to include channel dimension (1 for grayscale)
x_train = x_train.reshape((-1, 28, 28, 1))
x_test = x_test.reshape((-1, 28, 28, 1))

# One-hot encode the labels
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# Define the model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    MaxPooling2D((2, 2)),
    Conv2D(64, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Flatten(),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')  # 10 classes in MNIST
])

# Compile the model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
history = model.fit(x_train, y_train, epochs=10, validation_data=(x_test, y_test))

# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)
print(f'\nTest accuracy: {test_acc}')


11490434/11490434 [==============================] - 2s 0us/step
Epoch 1/10
1875/1875 [==============================] - 13s 5ms/step - loss: 0.1391 - accuracy: 0.9577 - val_loss: 0.0471 - val_accuracy: 0.9843
Epoch 2/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0476 - accuracy: 0.9856 - val_loss: 0.0343 - val_accuracy: 0.9876
Epoch 3/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0327 - accuracy: 0.9898 - val_loss: 0.0309 - val_accuracy: 0.9902
Epoch 4/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0236 - accuracy: 0.9924 - val_loss: 0.0289 - val_accuracy: 0.9906
Epoch 5/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0184 - accuracy: 0.9942 - val_loss: 0.0370 - val_accuracy: 0.9887
Epoch 6/10
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0135 - accuracy: 0.9957 - val_loss: 0.0254 - val_accuracy: 0.9915
Epoch 7/10
1875/1875 [==============================] - 8s 4ms/step - 

## Model Training and Evaluation Results

### Training Progress

- **Epochs**: The model was trained for 10 epochs. Each epoch represents a full pass through the training data.

- **Training Metrics**:
  - **Loss**: Measures the model’s prediction error. The training loss decreased from 0.1391 to 0.0069, indicating improved performance over time.
  - **Accuracy**: Represents the percentage of correct predictions. Training accuracy increased from 95.77% to 99.77%.

- **Validation Metrics**:
  - **Validation Loss**: Shows the model's performance on the validation set. It started at 0.0471 and slightly increased to 0.0431, suggesting good generalization.
  - **Validation Accuracy**: The percentage of correct predictions on the validation set. It ranged from 98.43% to 98.96%.

### Evaluation on Test Set

- **Test Accuracy**: 98.96%. This indicates that the model correctly classified 98.96% of the test images.

### Summary

The model demonstrates excellent performance with high accuracy on both training and test datasets. The loss values are low, indicating that the predictions are close to the true labels. Minor variations in validation accuracy suggest the model generalizes well to unseen data.

Overall, the model achieved very high accuracy in classifying MNIST digits, proving its effectiveness.
